### ISSUE TITLE
Keypoint detection on a single deforming object

### ISSUE THREAD

**Issue by dvthuytran** [ethz-mrl/SuperEvent#4]
Hi @yannickBurkhardt, thank you for a great work !

I tried your method to detect keypoints on my event stream of a butterfly. There are two cases in below:

<table border="1">
<caption><b>Case 1</b>: using <code>top_k=640</code> option for function <code>fast_nms</code>, then filter them by masking <code>pred["prob"] > 0.01</code></caption>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/ad6d6b7e-1395-480a-88d1-4e9a93a45634" width="360"><br>
      <sub>Image 1</sub>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/78808f60-a274-466e-931f-8fcd6f6aa085" width="360"><br>
      <sub>Image 2</sub>
    </td>
  </tr>
</table>

In Image 1, the keypoints are successfully detected, but I got no keypoints in Image 2. 
I thought it happens because the filter threshold is too high. So I removed the filtering process and got the following results:

<table border="1">
<caption><b>Case 2</b>: using <code>top_k=640</code> option for function <code>fast_nms</code></caption>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/5fb17c6c-43ce-4313-9419-5598de08188f" width="360"><br>
      <sub>Image 1</sub>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/f9afae98-2d52-441e-a29a-ac33d6ebc71c" width="360"><br>
      <sub>Image 2</sub>
    </td>
  </tr>
</table>

Image 2 still doesn't have any accurate keypoints since the number of key points is fixed to `top_k=640`.
Notably, without using filtering, it seems that the keypoints are _spatially replicated_ from the real keypoints on the butterfly. It might be caused by the pixel-level feature approach that is used in this method. 

Do you have any idea on why such spatial replication happened ? Also, I appreciate if you could give some suggestions on how to detect the keypoints more stably (for example in case of Image 2).

**Comment by yannickBurkhardt** [comment]
Hi @dvttran, thank you for your interest in our work.
The spatial patterns could be caused by the grid/ axis attention in the backbone. However, in my experience, the predicted values causing these patterns are very low, which is why I recommend to always use a threshold (`detection_threshold`) of at least 0.005 (the threshold and `top_k`-parameter might need some tuning depending on your application).

Your results on image 2 are not explained by that and surprising to me. Would you mind sharing the commands/ code you used and both MCTS of the butterfly?

**Comment by dvthuytran** [comment]
> Hi [@dvttran](https://github.com/dvttran), thank you for your interest in our work. The spatial patterns could be caused by the grid/ axis attention in the backbone. However, in my experience, the predicted values causing these patterns are very low, which is why I recommend to always use a threshold (`detection_threshold`) of at least 0.005 (the threshold and `top_k`-parameter might need some tuning depending on your application).
> 
> Your results on image 2 are not explained by that and surprising to me. Would you mind sharing the commands/ code you used and both MCTS of the butterfly?

Thank you so much, @yannickBurkhardt ! 

Fortunately, I found the issue while replying to you. When I trim the event data of Image 2, the result is pretty good as below. Before, I was running the model on multiple events, so maybe some issues on hardware/memory or data loader implementation that caused the inaccurate result.

<table border="1">
<caption>Using <code>top_k=640</code> option for function <code>fast_nms</code>, then filter them by masking <code>pred["prob"] > 0.01</code></caption>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/ad6d6b7e-1395-480a-88d1-4e9a93a45634" width="360"><br>
      <sub>Image 1</sub>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/037183d2-9eec-4fbe-b94b-498797bc331c" width="360"><br>
      <sub>Image 2</sub>
    </td>
  </tr>
</table>

For those who might be interested, here is the (incomplete) code :

```python
from models.super_event import SuperEvent
from util.visualization import ts2image
from einops import rearrange
import matplotlib.pyplot as plt

# Helper functions
def plot(mtcs, keypoints):
    mtcs = rearrange(mtcs, "C H W -> H W C")
    image = ts2image(mtcs)

    plt.imshow(image)
    plt.scatter(keypoints[..., 1], keypoints[..., 0], s=50, c=((0., 1., 0.),), marker='.')

# Path
model_config_path = "your/path/to/model/config/super_event.yaml"
model_path = "your/path/to/model/super_event_weights.pth"

# Load SuperEvent config
model_config = load_config(model_config_path)
height, width = 720, 1280
ts_shape = (height, width)
crop, crop_mask, cropped_shape = compute_crop_shape(ts_shape=ts_shape, config=model_config)
device = "cpu"
compile = False

# Load SuperEvent model
model = SuperEvent(model_config, tracing=compile)
state = torch.load(model_path, weights_only=True, map_location=device)
model.load_state_dict(state)
model.to(device).eval()

# Convert event data to MCTS
#     events: Array[(t, x, y, p)]
#     mcts: (C, H, W)
mcts = convert_event_to_mcts(events)

# Extract keypoints
top_k = np.max(ts_shape) // 2
with torch.inference_mode():
    preds = model(mcts)
    features_preds = extract_keypoints(preds, crop, model_config, compile, top_k=top_k)
    raw_keypoints = features_preds["keypoints"]

# Filter keypoints
threshold = 0.01
features_preds = filter_keypoints(features_preds, threshold)
keypoints = features_preds["keypoints"]

# Plot image with keypoints
plot(mcts, keypoints)
plt.show()
```
The helper functions `load_config`, `compute_crop_shape`, `extract_keypoints` and `filter_keypoints` are based on this [example](https://github.com/ethz-mrl/SuperEvent/blob/7d71ec52ecb1ab4fc5bcf6fba2e7d75d4ec0d078/evaluate_pose_estimation.py#L24).
